File: js/examples/live-trades.ts
import { MetaScalpClient, MetaScalpSocket } from '../src';

async function main() {
  // Discover MetaScalp
  const client = await MetaScalpClient.discover();
  const { connections } = await client.getConnections();
  const conn = connections[0];
  console.log(`Using connection [${conn.id}] ${conn.name}`);

  // Connect WebSocket
  const socket = await MetaScalpSocket.discover();
  console.log(`WebSocket connected on port ${socket.port}`);

  // Subscribe to trades
  const ticker = 'BTCUSDT';
  socket.subscribeTrades(conn.id, ticker);
  await socket.waitFor('trade_subscribed');
  console.log(`Subscribed to trades for ${ticker}`);

  // Listen for trade updates
  socket.on('trade_update', (data) => {
    for (const trade of data.trades) {
      const side = trade.side === 'Buy' ? '\x1b[32mBUY \x1b[0m' : '\x1b[31mSELL\x1b[0m';
      console.log(`${side} ${trade.price} x ${trade.size} @ ${trade.time}`);
    }
  });

  // Listen for errors
  socket.on('error', (data) => {
    console.error('Socket error:', data.error);
  });

  // Keep running
  console.log('Listening for trades... (Ctrl+C to stop)');
  process.on('SIGINT', () => {
    socket.unsubscribeTrades(conn.id, ticker);
    socket.disconnect();
    process.exit(0);
  });
}

main().catch(console.error);

File: python/examples/live_trades.py
"""Stream live trades from MetaScalp."""

import asyncio
from metascalp import MetaScalpClient, MetaScalpSocket


async def main():
    # Find connection
    client = await MetaScalpClient.discover()
    data = await client.get_connections()
    conn = data["connections"][0]
    await client.close()
    print(f"Using connection [{conn['id']}] {conn['name']}")

    # Connect WebSocket
    socket = await MetaScalpSocket.discover()
    print(f"WebSocket connected on port {socket.port}")

    # Register trade handler
    @socket.on("trade_update")
    def on_trade(data):
        for trade in data["trades"]:
            side = "\033[32mBUY \033[0m" if trade["side"] == "Buy" else "\033[31mSELL\033[0m"
            print(f"  {side} {trade['price']} x {trade['size']} @ {trade['time']}")

    @socket.on("trade_subscribed")
    def on_subscribed(data):
        print(f"Subscribed to trades for {data['ticker']}")

    # Subscribe
    ticker = "BTCUSDT"
    socket.subscribe_trades(conn["id"], ticker)

    print("Listening for trades... (Ctrl+C to stop)")
    await socket.listen_forever()


asyncio.run(main())
